presto-spi/src/main/java/com/facebook/presto/spi/block/ColumnarRow.java

```diff
@@ -31,27 +31,21 @@ public static ColumnarRow toColumnarRow(Block block)
             return toColumnarRow((RunLengthEncodedBlock) block);
         }
 
-        if (!(block instanceof AbstractArrayBlock)) {
+        if (!(block instanceof AbstractRowBlock)) {
             throw new IllegalArgumentException("Invalid row block");
         }
 
-        AbstractArrayBlock arrayBlock = (AbstractArrayBlock) block;
-        Block arrayBlockValues = arrayBlock.getValues();
-        if (!(arrayBlockValues instanceof AbstractInterleavedBlock)) {
-            throw new IllegalArgumentException("Invalid row block");
-        }
-        AbstractInterleavedBlock interleavedBlock = (AbstractInterleavedBlock) arrayBlockValues;
+        AbstractRowBlock rowBlock = (AbstractRowBlock) block;
 
         // get fields for visible region
-        int interleavedBlockOffset = 0;
-        int interleavedBlockLength = 0;
-        if (arrayBlock.getPositionCount() > 0) {
-            interleavedBlockOffset = arrayBlock.getOffset(0);
-            interleavedBlockLength = arrayBlock.getOffset(arrayBlock.getPositionCount()) - interleavedBlockOffset;
+        int firstRowPosition = rowBlock.getColumnBlockOffset(0);
+        int totalRowCount = rowBlock.getColumnBlockOffset(block.getPositionCount()) - firstRowPosition;
+        Block[] fieldBlocks = new Block[rowBlock.numColumns];
+        for (int i = 0; i < fieldBlocks.length; i++) {
+            fieldBlocks[i] = rowBlock.getColumnBlock(i).getRegion(firstRowPosition, totalRowCount);
         }
-        Block[] fields = interleavedBlock.computeSerializableSubBlocks(interleavedBlockOffset, interleavedBlockLength);
 
-        return new ColumnarRow(block, fields);
+        return new ColumnarRow(block, fieldBlocks);
```

Comment:
Looks like you intend to switch everything over at once. This is ok. I just want to check with you that this is your actual intention.